program-runtime/src/invoke_context.rs

```diff
@@ -65,31 +65,56 @@ pub trait Executor: Debug + Send + Sync {
 
 pub type Executors = HashMap<Pubkey, TransactionExecutor>;
 
-/// Tracks whether a given executor is "dirty" and needs to
-/// updated in the executors cache
+/// Tracks whether a given executor is "dirty" and needs to updated in the
+/// executors cache
 pub struct TransactionExecutor {
-    pub executor: Arc<dyn Executor>,
-    pub is_dirty: bool,
+    executor: Arc<dyn Executor>,
+    is_miss: bool,
+    is_updated: bool,
 }
 
 impl TransactionExecutor {
-    /// Wraps an executor and tracks that it doesn't need
-    /// to be updated in the executors cache.
-    pub fn cached(executor: Arc<dyn Executor>) -> Self {
+    /// Wraps an executor and tracks that it doesn't need to be updated in the
+    /// executors cache.
+    pub fn new_cached(executor: Arc<dyn Executor>) -> Self {
         Self {
             executor,
-            is_dirty: false,
+            is_miss: false,
+            is_updated: false,
         }
     }
 
-    /// Wraps an executor and tracks that it needs to be
-    /// updated in the executors cache.
-    pub fn dirty(executor: Arc<dyn Executor>) -> Self {
+    /// Wraps an executor and tracks that it needs to be updated in the
+    /// executors cache.
+    pub fn new_miss(executor: Arc<dyn Executor>) -> Self {
         Self {
             executor,
-            is_dirty: true,
+            is_miss: true,
+            is_updated: false,
         }
     }
+
+    /// Wraps an executor and tracks that it needs to be updated in the
+    /// executors cache only if the transaction succeeded.
+    pub fn new_updated(executor: Arc<dyn Executor>) -> Self {
+        Self {
+            executor,
+            is_miss: false,
+            is_updated: true,
+        }
+    }
+
+    pub fn is_dirty(&self, include_updates: bool) -> bool {
+        self.is_miss || (include_updates && self.is_updated)
+    }
+
+    pub fn get(&self) -> Arc<dyn Executor> {
+        self.executor.clone()
+    }
+
+    pub fn miss(&mut self) {
+        self.is_miss = true;
+    }
```

Comment:
This method is kinda dangerous because it could be used to make an "updated executor" into a cache miss and cause it to be considered dirty even when updates are not included. Can you add a `#[cfg(test)]` or rename to `set_miss_for_tests` or something?

Reply:
I tried adding `#[cfg(test)]` but that didn't work, maybe that only works locally.  I can rename to `for_tests`